### DISCUSSION TITLE
Recovering files uploaded using "Add files" that is overlapped by force push

### DISCUSSION THREAD

**Discussion by zakkutachibana** [community/community#58144] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

My team has been uploading some files on github repository, and then I force pushed files and now all files in the repository is gone. I see the commit changes and didn't see the uploaded files my team done using the "add file" feature. I revert the commit and the uploaded file using "add file" is still missing. Is there a way to see where these files located?

**Comment by spenserblack** [comment, marked as answer]
GitHub does keep a copy of commits, even if they've been "removed" from the repository. For example, I added a file, then I called `git reset --hard HEAD^` and `git push -f` to get rid of it. But I can still see the commit at https://github.com/spenserblack/test-repo/commit/6765ad8d8d352b34edf319ffd289415854ade09f (for now).
 
In my original local repository where I did the force-push, I can call `git checkout 6765ad8d8d352b34edf319ffd289415854ade09f` and see the file that I got rid of. This is because the commit still exists in that repo, even if it's inaccessible through the history. However, with a *new clone,* calling `git checkout 6765ad8d8d352b34edf319ffd289415854ade09f` results in a fatal error. So if you pulled down your teammate's changes *before* you force-pushed, you might be able to recover them if you can identify the commit hash.

You might also be able to fetch the files from the API, or at least find *where* you can fetch them. For example, https://api.github.com/repos/spenserblack/test-repo/commits/6765ad8d8d352b34edf319ffd289415854ade09f has the `"files"` key with a few helpful values, like the URL to the file's blob. You can view [the docs](https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28#get-a-commit) for more details.

******

Also, this isn't a solution to your immediate problem, but I highly recommend changing your workflow so that you're not all working on the same branch, and team members contribute via PRs to the main branch instead. Much easier to recover when you have a PR to reference, because you have a record of the commits that were made and the merge commit. Also, you can use a branch protection to block force pushing.

**Reply by zakkutachibana** [reply]
Thank you for the answer!
I actually managed to overcome this problem by going to the activity of the repository, and see the change history in the repo and managed to find the "overlapped" files.
Also yes! I have learnt my lesson and I will use branches for now on.
